File: CodeChef/JuneChallenge2020/ChefandIcecream/CHFICRM.py
import sys 

#   below 2 lines are for running code on your local machine 
#   Remove these 2 lines while using code online on codechef as they provide thier own input
sys.stdin = open("input.txt",'r')
sys.stdout = open("output.txt",'w')

# cook your dish here
t =int(input())
for j in range(t):
     n = int(input())
     arr = list(map(int,input().split()))
     first=0
     second=0
     #third=0
     flag=1
     for i in range(len(arr)):
         if arr[i]==5:
             first+=1
         elif arr[i]==10:
             second+=1
             if first>=1:
                 first-=1
             else:
                 flag=0
                 break
         else:
             if second>=1:
                 second-=1
             elif first>=2:
                 first-=2
             else:  
                 flag=0
                 break;
     if flag==0:
         print("NO")
     else:
         print("YES")

File: CodeChef/JuneChallenge2020/ChefandIcecream/CHFICRM.cpp
#include <iostream>
using namespace std;

int main() {
	// your code goes here
    //below 2 lines are for running code on your local machine 
    //Remove these 2 lines while using code online on codechef as they provide thier own input
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    
    int t;
    cin>>t;
    for(int i=0;i<t;i++)
    {
        int n;
        cin>>n;
        int first=0;
        int second=0;
        int third=0;
        int arr[n];
        for(int j=0;j<n;j++)
        {
            cin>>arr[j];
        }
        int flag=1;
        for(int k=0;k<n;k++)
        {
            if(arr[k]==5)
            {
                first++;
            }
            else if(arr[k]==10)
            {
                second++;

                if(first>=1)
                {
                    first--;
                }
                else
                {
                    flag=0;
                    break;
                }
                
            }
            else 
            {
                third++;
                if(second>=1)
                {
                    second--;
                }
                else if(first>=2)
                {
                    first=first-2;
                }
                else
                {
                    flag=0;
                    break;
                }
            }
        }
        if(flag==0)
            cout<<"NO"<<endl;
        else
            cout<<"YES"<<endl;
    }
	return 0;
}
